### PR TITLE
initial version

### PR DESCRIPTION
**Pull request by gaborzelei**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -25,5 +24,13 @@
       "Zap\\Routing\\": "src/",
       "Test\\": "tests/"
     }
-  }
+  },
+  "minimum-stability": "dev",
+  "repositories": [{
+    "type": "vcs",
+    "url": "https://github.com/php-zap/zap-common.git"
```
Comment (picahielos) [inline composer.json:31]: No need for this, they are in packagist (both).

File: src/ControllerDispatcher.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            throw $e;
+        }
+    }
+
+    /**
+     * Calls a controller based on current config
+     * @param array $params
+     * @return mixed
+     */
+    public function dispatchController(array $params = [])
```
Comment (picahielos) [inline src/ControllerDispatcher.php:47]: Do you think we could enforce to return `Response` class or `IResponse` or something similar?

File: src/Route.php [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        $this->userVariables = $userVariables;
+        return $this;
+    }
+
+    /**
+     * Executes route
+     * @return mixed                    @todo Change to Zap\Http\Response|array|bool|mixed
+     * @throws InvalidRouteException
+     */
+    public function process()
```
Comment (picahielos) [inline src/Route.php:186]: Same with the return of a response object (once it's done in `zap-http`).

File: tests/Routing/RouteTest.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        $this->assertInternalType('string', Route::define('/some/uri')->getMethod());
+    }
+
+    /**
+     * @covers Zap\Routing\Route::setMethod
+     */
+    public function testSetMethod()
+    {
+        $this->assertInstanceOf(Route::class,  Route::define('/some/uri')->setMethod('GET'));
```
Comment (picahielos) [inline tests/Routing/RouteTest.php:40]: In all those set tests I would also add an assertion making sure the getter is retuning the updated value.

File: src/Router.php [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        return $this;
+    }
+
+    /**
+     * Attempts to find a route based on current uri and request method
+     * @param string $uri
+     * @param string $method
+     * @return mixed|null
+     */
+    public function fetch(string $uri, string $method)
```
Comment (picahielos) [inline src/Router.php]: This could get as argument a `Request`, which contains both the uri and the method.

**Comment by picahielos** [comment]
I should add a `Response` object in `zap-http` before we go forward with this PR, so you can use it as return type and add more tests to see how a Request is resolved from a set of routes. I would try to write `Response` in a day or two. What do you think?

**Comment by gaborzelei** [comment]
The `Response` class is definitely needed, so that's a no-brainer. On the other hand this PR is not yet ready, I'm still working on it, so I'll give you a ping when it can be reviewed again. Your suggestions are in sync with what I still wanted to do in this bit of code, but another review will still be needed before
merge. Also, please let me know when `Request` and `Response` are both in `zap-http`.

Cheers :)

File: src/ControllerDispatcher.php (hunk 2 above)
Comment (gaborzelei) [inline src/ControllerDispatcher.php:47]: I'd rather go for something along the lines of what Laravel do, where they automatically generate a response from whatever the controller method returns.
E.g.: 
If return value is an `array` or an `object`, `Response` should render it into a `JSON` output.
If it's an actual `Response`, we just pass it on.
In any other cases, the output should be a standard `string` for now. (We can refine it later on, if the need arises.)

File: composer.json (hunk 1 above)
Comment (gaborzelei) [inline composer.json:31]: :+1:

File: tests/Routing/RouteTest.php (hunk 4 above)
Comment (gaborzelei) [inline tests/Routing/RouteTest.php:40]: :+1:

File: src/Router.php (hunk 5 above)
Comment (gaborzelei) [inline src/Router.php]: Yes, that is viable. I will need the `getUri()` and `getMethod()` getters in `Request` for that to work though.

**Pull request merged**